common/build/eslint-plugin-fluid/src/rules/no-unchecked-record-access.js

```diff
@@ -302,86 +301,150 @@ function isUndefinableIndexSignatureType(parserServices, node) {
+				return true;
+			}
+		}
+
+		// Handle presence checks
+		if (parent.type === "IfStatement" || parent.type === "ConditionalExpression") {
+			if (parent.test) {
+				if (nodesAreEquivalent(parent.test, node)) {
 					return true;
 				}
+				if (parent.test.type === "BinaryExpression") {
+					if (parent.test.operator === "in") {
+						const testBase = getBaseObject(parent.test.right);
+						if (baseObj === testBase) {
+							// Check if the else block assigns the key
+							const ifStatement = parent;
+							const elseBlock = ifStatement.alternate;
+							if (elseBlock) {
+								const keyNode = parent.test.left;
+								if (
+									checkElseBlockAssignsKey(elseBlock, testBase, keyNode, context)
+								) {
+									return true;
+								}
+							}
+							return true;
+						}
+					}
+					if (
+						(parent.test.operator === "!==" || parent.test.operator === "!=") &&
+						((nodesAreEquivalent(parent.test.left, node) &&
+							isUndefinedNode(parent.test.right)) ||
+							(nodesAreEquivalent(parent.test.right, node) &&
+								isUndefinedNode(parent.test.left)))
+					) {
+						return true;
+					}
+				}
 			}
 		}
 
-		// Handle "in" operator checks
-		if (
-			(current.parent.type === "ConditionalExpression" ||
-				current.parent.type === "IfStatement") &&
-			current.parent.test?.type === "BinaryExpression" &&
-			current.parent.test.operator === "in"
-		) {
-			const test = current.parent.test;
-			const base = getBaseObject(node);
-			const testBase = getBaseObject(test.right);
-			if (base === testBase) {
-				return true;
+		// Additional check for 'in' checks in the same scope
+		const containingBlock = findContainingBlock(current);
+		if (containingBlock) {
+			for (const statement of containingBlock.body) {
+				if (statement.range[0] > current.range[0]) break; // Only check statements before current node
+				if (
+					statement.type === "IfStatement" &&
+					statement.test?.type === "BinaryExpression" &&
+					statement.test.operator === "in"
+				) {
+					const testBase = getBaseObject(statement.test.right);
+					const testKey = statement.test.left;
+					if (testBase === baseObj && nodesAreEquivalent(testKey, currentKeyNode)) {
+						const elseBlock = statement.alternate;
+						if (
+							elseBlock &&
+							checkElseBlockAssignsKey(elseBlock, testBase, testKey, context)
+						) {
+							return true;
+						}
+					}
+				}
 			}
 		}
 
-		current = current.parent;
+		current = parent;
 	}
 
-	if (
-		node.parent.type === "BinaryExpression" &&
-		node.parent.operator === "in" &&
-		node.parent.left === node
-	) {
-		return true;
+	return false;
+}
+
+function isUndefinedNode(node) {
+	return (
+		(node.type === "Identifier" && node.name === "undefined") ||
+		(node.type === "UnaryExpression" &&
+			node.operator === "void" &&
+			node.argument?.type === "Literal" &&
+			node.argument.value === 0)
+	);
+}
+
+function nodesAreEquivalent(a, b) {
```

Comment:
Nit: docs here would probably be good. How conservative is this check? Is this likely to report false-negatives?

Reply:
+1. I was a bit concerned about the `!a || !b` check being too vague (empty strings coalescing to false, things like that), but it seems this function explicitly expects AST nodes for both parameters, right? Is there anything we can do to be a bit more strict about that? Validating that the parameters are indeed that and not something else.

Reply:
Updated the check to use this helper function instead
```js
function isNode(node) {
	return (
		node !== null &&
		node !== undefined &&
		typeof node === "object" &&
		"type" in node &&
		typeof node.type === "string" &&
		"parent" in node &&
		(node.type === "Identifier" ||
			node.type === "Literal" ||
			node.type === "MemberExpression" ||
			node.type === "BinaryExpression")
	);
}
```